test/WebSites/TempDataWebSite/Controllers/HomeController.cs

```diff
@@ -41,5 +43,49 @@ public IActionResult PeekTempData()
             var peekValue = TempData.Peek("key");
             return Content(peekValue.ToString());
         }
+
+        public IActionResult SetTempDataMultiple(
+            string value,
+            int intValue,
+            IList<string> listValues,
+            DateTime datetimeValue,
+            Guid guidValue)
+        {
+            TempData["key1"] = value;
+            TempData["key2"] = intValue;
+            TempData["key3"] = listValues;
+            TempData["key4"] = datetimeValue;
+            TempData["key5"] = guidValue;
+            return RedirectToAction("GetTempDataMultiple");
+        }
+
+        public string GetTempDataMultiple()
+        {
+            var value1 = TempData["key1"].ToString();
+            var value2 = Convert.ToInt32(TempData["key2"]);
+            var value3 = (IList<string>)TempData["key3"];
+            var value4 = (DateTime)TempData["key4"];
+            var value5 = (Guid)TempData["key5"];
+            return $"{value1} {value2.ToString()} {value3.Count.ToString()} {value4.ToString()} {value5.ToString()}";
+        }
+
+        public string SetTempDataInvalidType()
```

Comment:
This is being used in this test https://github.com/aspnet/Mvc/pull/2317/files#diff-c244bc4de89efad52c8eb4342c2ea7c9R192 to verify it throws for types that can't be serialized.